contracts/stake-cw20-external-rewards/src/contract.rs

```diff
@@ -0,0 +1,1324 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    from_binary, to_binary, Addr, BankMsg, Binary, Coin, CosmosMsg, Deps, DepsMut, Env,
+    MessageInfo, QuerierWrapper, Response, StdResult, Timestamp, Uint128, WasmMsg,
+};
+use cw20::{Cw20ExecuteMsg, Cw20ReceiveMsg, Denom};
+use std::cmp::min;
+
+use crate::error::ContractError;
+use crate::error::ContractError::RewardsNotStarted;
+use crate::msg::{
+    ClaimableRewardsResponse, ExecuteMsg, InfoResponse, InstantiateMsg, QueryMsg, ReceiveMsg,
+};
+use crate::state::{Config, LastClaim, CONFIG, LAST_CLAIMED};
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    validate_instantiate_msg(&env, &msg)?;
+    let config = Config {
+        start_block: msg.start_block,
```

Comment:
What's the rationale for having the user pass in start_block, end_block, and payment per block? It would be more elegant to accept a duration and an amount, and then calculate payment per unit time within the contract. I do something similar [here](https://github.com/CosmWasm/cw-tokens/blob/4ca9307f10b94f545427963c78bf10a0ef9c7641/contracts/cw20-streams/src/contract.rs#L61).

Reply:
I like that the parameters are explicit and the instatiator is required to double check everything will function as expected.